### PR TITLE
[FLINK-21911][table] Add built-in greatest/least functions support

### PR DESCRIPTION
**Pull request by snuyanzin**
## What is the purpose of the change

The PR implements three functions from the list mentioned in FLINK-21911 : `GREATEST`, `LEAST`

## Brief change log

Implementation, tests and docs for `GREATEST`, `LEAST`

## Verifying this change

- `MiscFunctionsITCase`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs / JavaDocs

### CODE REVIEW

**Comment by wuchong** [comment]
Hi @snuyanzin , is this pull request ready to be reviewed? 

Besides, implementating a built-in function should be simple using the `BuiltInScalarFunction`, you can learn how to implement one from FLINK-20522.

**Comment by snuyanzin** [comment]
Hi @wuchong yes it is ready.

Yes, I know about FLINK-20522 however the issue is that new-stack built-in functions do not support compare operations which are required here....

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/ScalarOperatorGens.scala [hunk 1]
```diff
@@ -1756,6 +1756,56 @@ object ScalarOperatorGens {
...
+         |     $nullTerm = true;
+         |   } else {
+         |     int compareResult = $result.compareTo($cur);
+         |     if ($greatest && compareResult < 0 || compareResult > 0 && !$greatest) {
+         |       $result = $cur;
+         |     }
+         |   }
+         | }
+       """.stripMargin
+    }.mkString("\n")
```
Comment (leonardBang) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/ScalarOperatorGens.scala:1795]: Looks we didn't use `idx` and generated code for each input element,  could we optimize to one loop?

File: flink-python/pyflink/table/expressions.py [hunk 2]
```diff
@@ -505,6 +505,34 @@ def if_then_else(condition: Union[bool, Expression[bool]], if_true, if_false) ->
...
+
+    e.g. least(1, 2, 3) leads to 1
+
+    .. seealso:: :func:`~pyflink.table.expressions.greatest`
+    """
+    gateway = get_gateway()
+    tail = to_jarray(gateway.jvm.Object, [_get_java_expression(t) for t in tail])
+    return _binary_op("least", head, tail)
+
+
```
Comment (leonardBang) [inline flink-python/pyflink/table/expressions.py:535]: ditto, python expression can also call SQL builtin function.

**Review by leonardBang — COMMENTED** [review]
Thanks @snuyanzin for the contribution, the PR is in good shape, I left some minor comments.

**Comment by wuchong** [comment]
Hi @snuyanzin , do you have time to update the  PR? Otherwise, we can pick it up and help to polish it.

**Comment by snuyanzin** [comment]
Hi @wuchong , unfortunately the nearest date when I can have a look here is second part of June...
So feel free to continue in case you can not wait

**Comment by leonardBang** [comment]
@snuyanzin I'll take over this one and keep your contribution.

**Comment by wuchong** [comment]
Merged in #16020.

**Pull request closed without merging**